Hexastore.Web/Queue/QueueWriter.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Hexastore.Web.EventHubs;
+using Microsoft.Extensions.Logging;
+
+namespace Hexastore.Web.Queue
+{
+    public class QueueWriter : IDisposable
+    {
+        private readonly BlockingCollection<StoreEvent> _queue;
+        private readonly EventReceiver _eventReceiver;
+        private readonly ILogger _logger;
+        private readonly Task _task;
+        private readonly CancellationTokenSource _cts = new CancellationTokenSource();
+
+        public QueueWriter(EventReceiver eventReceiver, ILogger logger)
+        {
+            _queue = new BlockingCollection<StoreEvent>();
+            _eventReceiver = eventReceiver;
+            _logger = logger;
+            _task = StartReader();
+            _task.Start();
+        }
+
+        public int Length => _queue.Count;
+
+        public void Dispose()
+        {
+            _cts.Cancel();
+            _queue.Dispose();
+        }
+
+        public void Send(StoreEvent storeEvent)
+        {
+            _queue.Add(storeEvent);
+        }
+
+        private Task StartReader()
+        {
+            return new Task(() => {
+                while (!_cts.IsCancellationRequested) {
+                    try {
+                        var storeEvent = _queue.Take();
```

Comment:
I think a ConcurrentQueue would be simpler than a BlockingCollection here. 

On TryDequeue if the queue is empty the thread should sleep with Task.Delay, then it would give up the thread instead of having 32 threads continually running context switches with BlockingCollection.Take even if no work is happening.

Reply:
I think BlockingCollection Take is going to change the state of the Task and it is not a busy wait. I have used it before and it works fairly well. With ConcurrentQueue we would have to manually write the sleep/signaling logic. BlockingCollection does that for us. Having 32 Tasks in the system does not add a significant overhead.